[universal] - Issue universal config change for non-root default codespace user and installing google chrome browser reuse sandbox to run puppeteer cli in universal image

Kaniska244 opened the pull request:
**Ref**: https://github.com/devcontainers/internal/issues/247 
               https://github.com/devcontainers/internal/issues/249 

**Dev container name:**

Universal

**Description:**

We can determine UID of host machine via id -u & id -g and pass it to docker while building the image. Modified devcontainer.json as below resolved the issue.

**Changelog:**

For Universal UID permission issue

- Updated GitHub workflows to determine UID of host machine via id -u & id -g and pass it to GITHUB_ENV & updated devcontainer.json file to read the UID & GID from localEnv.

For puppeteer library issue

- Updated setup-user local feature to install chrome & also changed the devcontainer.json to populate a specific env variable in the remoteEnv, required for puppeteer library.

**Checklist:**

- Checked that applied changes work as expected.

Kaniska244 commented:
Hi @ddoyle2017 ,

Tested another workaround for this instead of forcing the UID 1001 as that was increasing the size of the image considerably. 

- Setting "updateRemoteUserUID" flag to false so that UID & GID aren't replaced, so that the UID & GID of codespace user remains 1000.

- Adding "postStartCommand" to change the ownership of /workspaces/images folder to the same as of the codespace user, not using "postCreateCommand" option as its already used in other sub-features in universal image. This folder is directly mounted from the local machine containing the images repo sources.

- Tests are fine with this. Also tested locally in a codespace along with the devcontainer CI pipeline.

Would you kindly review the same. Please let me know in case of any concern.

With Regards,
Kaniska

ddoyle2017 reviewed (changes requested):
Almost there! Can we revert the puppeteer `--no-sandbox` changes and rerun the Actions?

Kaniska244 commented:
Hi @ddoyle2017 ,

Test one fix for the puppeteer cli issue. Uploaded the fix. Would you kindly review the same & let me know in case of any concern. 
Please note the change is only in the test script as the purpose of the test is to demonstrate that its possible to run a puppeteer node app in the universal container.

PFB the details:

- The actual issue was that there was no sandbox available to launch the browser when runner image is on ubuntu-24.04.
- Downloaded google chrome amd64-deb variant latest stable version & installed the same in the container. 
- Chrome is shipped with a default sandbox, copied the same & populated this CHROME_DEVEL_SANDBOX variable which is used by the puppeteer by default.
- All tests are fine with this & this appears to be the safest approach compared to the other approaches which would require to launch the browser directly in the container without the sandbox or modifying the kernel to allow unprivileged access. 

Ref - https://github.com/puppeteer/puppeteer/blob/88ef09c6ccc27d024972725c728a9db50f010f36/docs/troubleshooting.md#setting-up-chrome-linux-sandbox 

With Regards,
Kaniska Sengupta

ddoyle2017 reviewed (changes requested):
I think this is a reasonable work around for the Puppeteer issues with Ubuntu 24.04! Can you add a comment above both your changes to explain what issue they're solving?

eljog commented on src/universal/.devcontainer/devcontainer.json:
```diff
@@ -103,7 +103,8 @@
     "containerUser": "codespace",
-
+    "updateRemoteUserUID": false,
+    "postStartCommand": "sudo chown -R 1000:1000 /workspaces/images/",
```
May I know why this need to be a `postStartCommand`?
`postStartCommand` re runs on every time container is restarted. We need to do this work only once

Kaniska244 replied:
Hi @eljog ,
Indeed it would be executed during every restart. Initially I saw that as part of configurations in one of the features of universal image, git-lfs has already got postCreateCommand used in devcontainer-feature.json. So I avoided the postCreateCommand option & opted for postStartCommand. 
However, I have checked now that for a devcontainer configuration if the same lifecycle hook such as postCreateCommand is used in both devcontainer-feature.json & devcontainer.json, then the commands provided in the devcontainer-feature.json would be executed first followed by the commands in devcontainer.json. So I changed that & tested the same now. Kindly let me know in case of any further concerns on this.

With Regards,
Kaniska

Kaniska244 replied:
The approach is now changed. Resolving the conversation.

eljog commented on src/universal/test-project/test.sh:
```diff
@@ -133,6 +133,19 @@ check "zsh" zsh --version
+export CHROME_DEVEL_SANDBOX=/usr/local/sbin/chrome-devel-sandbox
+cd /workspaces/images/src/universal/test-project/
 yarn
 check "run-puppeteer" node puppeteer.js
```
Will this command work on the container when the user runs it?
Or is it passing here because the additional setup done above, before running the check?

Kaniska244 replied:
Hi @eljog ,

My idea was that the main objective behind this particular test "run-puppeteer" was to demonstrate that puppeteer cli library can be used in a node app & the same cane be executed in the universal image. So I changed it in the test.sh alone considering that & also due to the fact that as part of the solution I am downloading & installing google chrome in the image which increases the image size if done as part of the build. 

Now I have changed that. Following points are needed to be considered for this change.

- Made change in Dockerfile to install three more libraries which are required to install chrome.
- Made change in install.sh of the local feature setup-user to install chrome & placing the sandbox in the desired location. Please let me know if I should introduce a new local feature for this instead.
- Made change in devcontainer.json to configure "remoteEnv" tag for the CHROME_DEVEL_SANDBOX variable with the sandbox location which is used by the puppeteer library while launching the browser.
- Size of the image has now increased considerably due to the chrome installation which is still within the configured threshold but only by about 64 MB. With the previous change the gap was about 422 MB. 

Kindly let me know in case of any concerns on this.

With Regards,
Kaniska Sengupta

eljog replied:
My concern is more about the user experience here.
If we don't install chrome on the image, will the puppeteer scenarios start breaking for them from the new version?
Are the users expected to manually install the setup each time, to get things working? Not sure if that's the expectation on using universal images.

Tagging other maintainers on their thoughts. @chrmarti @bamurtaugh

Kaniska244 replied:
> My concern is more about the user experience here. If we don't install chrome on the image, will the puppeteer scenarios start breaking for them from the new version? Are the users expected to manually install the setup each time, to get things working? Not sure if that's the expectation on using universal images.
> 
> Tagging other maintainers for their thoughts. @chrmarti @bamurtaugh

Hi @eljog ,

Actually right now on the image built on top of ubuntu-24.04 runner image in CI, no sandbox is available to reuse for  puppeteer node app & due to enhanced security feature of ubuntu-24.04, its not possible to create sandbox without enabling user namespace cloning which is still has some vulnerabilities. 
Ref:- https://github.com/puppeteer/puppeteer/blob/88ef09c6ccc27d024972725c728a9db50f010f36/docs/troubleshooting.md#setting-up-chrome-linux-sandbox 

https://github.com/puppeteer/puppeteer/issues/12818#issuecomment-2415915338 

With Regards,
Kaniska

ddoyle2017 replied:
Closing out this conversation since its outdated and the PR has 2 approvals 👍

ddoyle2017 commented:
hey @Kaniska244, so @eljog and I had a discussion about how to move forward with the release and I think we need to reorganize your changes for Puppeteer.

Currently, your Chrome workaround works for our CI test, but Puppeteer will still be broken for users of this image. We should apply your updates to the image as a whole (not just in `test.sh`), then see if that fixes the issue 👍

Kaniska244 commented:
> hey @Kaniska244, so @eljog and I had a discussion about how to move forward with the release and I think we need to reorganize your changes for Puppeteer.
> 
> Currently, your Chrome workaround works for our CI test, but Puppeteer will still be broken for users of this image. We should apply your updates to the image as a whole (not just in `test.sh`), then see if that fixes the issue 👍

Hi @ddoyle2017 ,

I have made the changes according to the comments & all tests are fine with this. Would you kindly review the latest change.
PFB the details of the change.

- Changed the postStartCommand option to postCreateCommand as per the review comments.
-  Made change in Dockerfile to install three more libraries which are required to install chrome.
- Made change in install.sh of the local feature setup-user to install chrome & placing the sandbox in the desired location. Please let me know if I should introduce a new local feature for this instead.
- Made change in devcontainer.json to configure "remoteEnv" tag for the CHROME_DEVEL_SANDBOX variable with the sandbox location which is used by the puppeteer library while launching the browser.
- Size of the image has now increased considerably due to the chrome installation which is still within the configured threshold but only by about 64 MB. With the previous change the gap was about 422 MB.
- Added comments on top of the changes for more clarity.

With Regards,
Kaniska

Kaniska244 commented:
Hi @ddoyle2017 , Hi @eljog ,

I have made further changes on the solution of the UID issue after consulting with @Mathiyarasy in order to achieve more clean, error-proof solution. Would you kindly review the same.  

PFB the details of the changes:

- Changed the GitHub action workflows(universal smoke test & all the build and push workflows) to copy the UID & GID of the host & set it as environment variable.

- Changed the devcontainer.json to read the environment variables for UID & GID instead of hardcoding them to 1000.

- In case if the environment variables aren't set, the UID & GID are defaulted to 'automatic' option in common-utils feature which creates the user & group with the default available UID & GID as 1000. The build doesn't fail in that case as I tested in my local where the build was manually triggered, not by workflow.

- Made changes in all of the build & push workflows to fetch the UID/ GID and set the environment variables. However not able to test them independently in the forked repo so far. Same change works fine for smoke test universal workflow.

- The image size is also back within the configured threshold with this change.

- Also kindly note that jekyll smoke test is failing but I see it has already been failing with the same, so not related to this.

Please let me know in case of further concerns with this latest change.

With Regards,
Kaniska

ddoyle2017 commented on src/universal/test-project/test.sh:
@Kaniska244 can you walk me through why this file was added?

Kaniska244 replied:
Hi @ddoyle2017 ,

There's actually no change in that test.sh file. There were changes in that file which were made during the course of this PR. But they are all removed now. Perhaps some extra tab or enter might be left over there & that's why still showing the file among the changed files. 

With Regards,
Kaniska

ddoyle2017 reviewed (dismissed):
Changes look good 🚀 I think this approach makes sense, based on [Option 3](https://chromium.googlesource.com/chromium/src/+/main/docs/security/apparmor-userns-restrictions.md#option-3_the-safest-way) of Chromium's recommended AppArmor workarounds

chrmarti commented on src/universal/.devcontainer/devcontainer.json:
```diff
@@ -103,7 +104,11 @@
-
+	// This variable is set to ensure puppeteer library gets the sandbox location by default.
+	// Ref:- https://github.com/devcontainers/internal/issues/249
+    "remoteEnv": {
```
Not sure this matters: Could use `"containerEnv"` to ensure this is also set when connecting using SSH (which 3rd party IDEs do?).

Kaniska244 replied:
Hi @chrmarti ,

Thanks a lot for the review. I have now changed it to "containerEnv". Kindly let me know in case of any firther concern with this.

With Regards,
Kaniska

chrmarti commented on src/universal/.devcontainer/devcontainer.json:
```diff
@@ -6,8 +6,9 @@
-            "userUid": "1000",
-            "userGid": "1000"
+			//Getting the UID & GID values of the host machine via GitHub actions
+            "userUid": "${localEnv:USER_UID}",
```
Do we know why `updateRemoteUserUID` doesn't work? This change makes the universal image follow the GitHub agent's UID:GID which seems unexpected, having `updateRemoteUserUID` handle this transparently would keep these decoupled.

Kaniska244 replied:
Hi @chrmarti ,

Please find below the explanations for the same. Kindly let me know in case of any concern.

- There has been a recent change in GitHub action runner side such that for larger runners the UID of the base user in pipeline "runner", is changed from 1000 to 1001 to align the same with standard runners. Ref: https://github.com/actions/runner-images/issues/10936

- During the build as part of local feature setup-user for the universal image (https://github.com/devcontainers/images/blob/cd4daf6c6b692cea79c83ff26c60bca44f51e8f7/src/universal/.devcontainer/local-features/setup-user/install.sh#L63C1-L90C49), certain installation paths are given the ownership of the codespace user. 

- After the build is done, the new UID & GID of the local user "runner" in the CI pipeline are getting copied to the "codespace" user inside the universal container as part of updateRemoteUserUID logic in the smoke-test-universal workflow https://github.com/devcontainers/images/blob/main/.github/workflows/smoke-universal.yaml

- When the UID & GID were overwritten due to the recent change in GitHub actions resulting in conflict  leaving certain files & folders inside the container with UID & GID as 1000 (these are the same set of files & directories for which the ownership was earlier changed as part of the setup-user local feature as mentioned above) which leads to the permission denied error faced in the smoke test in the step .

- Therefore as the solution we choose to follow UID & GID of the host for universal image build & the smoke test.

**Permission snapshot of dotnet directory:**

```
/home/codespace/.dotnet
total 152K
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:21 .
drwxr-xr-x 1 root      root      4.0K Jan 19 10:21 ..
-rw-rw-r-- 1 codespace codespace  220 Feb 25  2020 .bash_logout
-rw-rw-r-- 1 codespace codespace 5.8K Jan 19 09:50 .bashrc
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:21 .cache
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 09:53 .conda
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:21 .config
drwxr-sr-x 2 codespace codespace 4.0K Jan 19 10:21 .devcontainer
lrwxrwxrwx 1 codespace codespace   17 Jan 19 10:17 .dotnet -> /usr/share/dotnet
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:17 .hugo
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:16 .jupyter
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:16 .local
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:17 .maven
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:08 .minikube
drwxr-sr-x 3 codespace codespace 4.0K Jan 19 10:21 .npm
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 09:53 .nvs
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:08 .oh-my-zsh
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:17 .php
-rw-rw-r-- 1 codespace codespace  807 Feb 25  2020 .profile
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:17 .python
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:21 .rbenv
drwxrwsr-x 1 codespace codespace 4.0K Jan 19 10:17 .ruby
-rw-r--r-- 1 codespace codespace   39 Jan 19 10:21 .rvmrc
drwxr-sr-x 3 codespace codespace 4.0K Jan 19 10:21 .yarn
-rw-rw-r-- 1 codespace codespace   22 Jan 19 09:50 .zprofile
-rw-rw-r-- 1 codespace codespace 4.0K Jan 19 09:50 .zshrc
lrwxrwxrwx 1 codespace codespace   33 Jan 19 10:17 java -> /usr/local/sdkman/candidates/java
lrwxrwxrwx 1 codespace codespace   20 Jan 19 10:17 nvm -> /usr/local/share/nvm

/usr/share/dotnet
total 236K
drwxrwsr-x 1 1000 1000 4.0K Jan 19 10:10 .
drwxr-xr-x 1 root root 4.0K Jan 19 10:08 ..
-rw-rw-r-- 1 1000 1000 1.1K Jan 19 10:10 LICENSE.txt
-rw-rw-r-- 1 1000 1000  93K Jan 19 10:10 ThirdPartyNotices.txt
-rwxrwxr-x 1 1000 1000  67K Jan 19 10:10 dotnet
drwxrwxr-x 1 1000 1000 4.0K Jan 19 09:53 host
drwxrwxr-x 1 1000 1000 4.0K Jan 19 10:10 metadata
drwxrwxr-x 1 1000 1000 4.0K Jan 19 09:53 packs
drwxrwxr-x 1 1000 1000 4.0K Jan 19 10:11 sdk
drwxrwxr-x 1 1000 1000 4.0K Jan 19 09:53 sdk-manifests
drwxrwxr-x 1 1000 1000 4.0K Jan 19 09:53 shared
drwxrwxr-x 1 1000 1000 4.0K Jan 19 10:11 templates
```

**Error snapshot in smoke test:**

```
🧪 Testing oryx-install-dotnet-2.1
Operation performed by Microsoft Oryx, https://github.com/Microsoft/Oryx
You can report issues at https://github.com/Microsoft/Oryx/issues

Oryx Version: 0.2.0.0+cbef87d362c06307cbc86ede8a0bc278e5194445, Commit: , ReleaseTagName: 

Skipping platform detection since '--skip-detection' switch was used...
Version '2.1.30' of platform 'dotnet' is not installed. Generating script to install it...

Setting up environment...


Downloading and extracting 'dotnet' version '2.1.818' to '/opt/dotnet/2.1.818'...
Detected image debian flavor: focal-scm.
Downloaded in 8 sec(s).
Verifying checksum...
Extracting contents...
performing sha512 checksum for: dotnet...
Done in 11 sec(s).

image detector file exists, platform is dotnet..
image detector file exists, image is vso based..
Symlink directory name: codespace
Error: cp: cannot remove '/home/codespace/.dotnet/dotnet': Permission denied
❌ oryx-install-dotnet-2.1 check failed.
```

Kaniska244 replied:
Hi @chrmarti ,

Indeed, if I set this option update-remote-user-uid-default to 'never', certain test steps in universal smoke test workflow starts failing. All these failing steps require write permission inside the workspaces/images directory which is locally mounted from the runner machine & currently having the same ownership as that of the 'runner' user from runner machine (uid 1001 & gid 118). 
We had in fact tried the same earlier globally for the universal image by setting the updateRemoteUserUID flag to false in the devcontainer.json file coupled with workaround (a dirty one perhaps) to provide localized permission to the /workspaces/images directory by setting postCreateCommand with the following for the /workspaces/images/ directory.
```
"postCreateCommand": "/bin/sh -c 'set -e && if [ -d \"/workspaces/images/\” ]; then echo \"inside\" && sudo chown -R 1000:1000 /workspaces/images/; fi'"
```
Test failure snapshot with update-remote-user-uid-default to 'never' without any other workaround:

```
2025-02-18T06:46:18.2588218Z #
2025-02-18T06:46:18.2588600Z #     $ conda deactivate
2025-02-18T06:46:18.2588806Z 
2025-02-18T06:46:18.3515258Z ✅  Passed!
2025-02-18T06:46:18.3515533Z 
2025-02-18T06:46:18.3516168Z 💥  Failed tests: run-puppeteer oryx-build-node-projectr oryx-build-java-project oryx-build-python
2025-02-18T06:46:18.3651642Z Error occurred at line 15 with exit code 1 in command devcontainer
2025-02-18T06:46:18.3678157Z ##[error]Process completed with exit code 1.
```

```
2025-02-18T06:43:05.8397644Z .githubpreview.dev,.preview.app.github.dev,.app.github.dev
2025-02-18T06:43:05.8400353Z check the env variable.
2025-02-18T06:43:05.8400770Z /usr/local/sbin/chrome-devel-sandbox
2025-02-18T06:43:05.9685533Z yarn install v1.22.22
2025-02-18T06:43:05.9773610Z warning package.json: No license field
2025-02-18T06:43:05.9996384Z warning No license field
2025-02-18T06:43:06.0006247Z [1/4] Resolving packages...
2025-02-18T06:43:06.0336211Z [2/4] Fetching packages...
2025-02-18T06:43:06.8173261Z [3/4] Linking dependencies...
2025-02-18T06:43:06.8556147Z error Error: EACCES: permission denied, mkdir '/workspaces/images/src/universal/test-project/node_modules'
2025-02-18T06:43:06.8557062Z info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
2025-02-18T06:43:06.8721084Z 
2025-02-18T06:43:06.8721781Z 🧪 Testing run-puppeteer
2025-02-18T06:43:06.8959460Z node:internal/modules/cjs/loader:1228
2025-02-18T06:43:06.8959918Z   throw err;
2025-02-18T06:43:06.8960203Z   ^
2025-02-18T06:43:06.8960357Z 
2025-02-18T06:43:06.8960508Z Error: Cannot find module 'puppeteer'
2025-02-18T06:43:06.8960909Z Require stack:
2025-02-18T06:43:06.8961348Z - /workspaces/images/src/universal/test-project/puppeteer.js
2025-02-18T06:43:06.8962041Z     at Module._resolveFilename (node:internal/modules/cjs/loader:1225:15)
2025-02-18T06:43:06.8962982Z     at Module._load (node:internal/modules/cjs/loader:1051:27)
2025-02-18T06:43:06.8963602Z     at Module.require (node:internal/modules/cjs/loader:1311:19)
2025-02-18T06:43:06.8964517Z     at require (node:internal/modules/helpers:179:18)
2025-02-18T06:43:06.8965293Z     at Object.<anonymous> (/workspaces/images/src/universal/test-project/puppeteer.js:1:19)
2025-02-18T06:43:06.8966102Z     at Module._compile (node:internal/modules/cjs/loader:1469:14)
2025-02-18T06:43:06.8966782Z     at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
2025-02-18T06:43:06.8967445Z     at Module.load (node:internal/modules/cjs/loader:1288:32)
2025-02-18T06:43:06.8968048Z     at Module._load (node:internal/modules/cjs/loader:1104:12)
2025-02-18T06:43:06.8968952Z     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:173:12) {
2025-02-18T06:43:06.8969664Z   code: 'MODULE_NOT_FOUND',
2025-02-18T06:43:06.8970265Z   requireStack: [ '/workspaces/images/src/universal/test-project/puppeteer.js' ]
2025-02-18T06:43:06.8970881Z }
2025-02-18T06:43:06.8971027Z 
2025-02-18T06:43:06.8971155Z Node.js v20.18.3
2025-02-18T06:43:06.8985025Z ❌ run-puppeteer check failed.
```

```
2025-02-18T06:43:07.0138058Z 🧪 Testing oryx-build-node-projectr
2025-02-18T06:43:07.2797493Z Operation performed by Microsoft Oryx, https://github.com/Microsoft/Oryx
2025-02-18T06:43:07.2798316Z You can report issues at https://github.com/Microsoft/Oryx/issues
2025-02-18T06:43:07.2798604Z 
2025-02-18T06:43:07.2798890Z Oryx Version: 0.2.0.0+d363b83aad8365d7e91cc246d6dfb3ddad209ddb, Commit: , ReleaseTagName: 
2025-02-18T06:43:07.2799357Z 
2025-02-18T06:43:07.2801543Z Build Operation ID: 42b639716a241fb0
2025-02-18T06:43:07.2802050Z OS Type           : focal-scm
2025-02-18T06:43:07.2802746Z Image Type        : vso-focal
2025-02-18T06:43:07.2803008Z 
2025-02-18T06:43:07.2926507Z Detecting platforms...
2025-02-18T06:43:07.8114951Z Detected following platforms:
2025-02-18T06:43:07.8115522Z   nodejs: 16.20.2
2025-02-18T06:43:07.8145213Z Version '16.20.2' of platform 'nodejs' is not installed. Generating script to install it...
2025-02-18T06:43:07.9100146Z 
2025-02-18T06:43:07.9164094Z 
2025-02-18T06:43:07.9200891Z Source directory     : /workspaces/images/src/universal/test-project/sample/node
2025-02-18T06:43:07.9201812Z Destination directory: /workspaces/images/src/universal/test-project/sample/node
2025-02-18T06:43:07.9202332Z 
2025-02-18T06:43:07.9202337Z 
2025-02-18T06:43:07.9202554Z Downloading and extracting 'nodejs' version '16.20.2' to '/opt/nodejs/16.20.2'...
2025-02-18T06:43:07.9216751Z Detected image debian flavor: focal-scm.
2025-02-18T06:43:08.2267482Z Downloaded in 1 sec(s).
2025-02-18T06:43:08.2268493Z Verifying checksum...
2025-02-18T06:43:08.2323429Z Extracting contents...
2025-02-18T06:43:08.9540824Z performing sha512 checksum for: nodejs...
2025-02-18T06:43:09.0548912Z Done in 2 sec(s).
2025-02-18T06:43:09.0550674Z 
2025-02-18T06:43:09.0564767Z image detector file exists, platform is nodejs..
2025-02-18T06:43:09.1188262Z Removing existing manifest file
2025-02-18T06:43:09.1201629Z Creating directory for command manifest file if it does not exist
2025-02-18T06:43:09.1232057Z Creating a manifest file...
2025-02-18T06:43:09.1236822Z /tmp/BuildScriptGenerator/587f0f7905d84720a3692db17e4de32f/build.sh: line 262: /workspaces/images/src/universal/test-project/sample/node/oryx-build-commands.txt: Permission denied
2025-02-18T06:43:09.1526633Z ❌ oryx-build-node-projectr check failed.
```

```
2025-02-18T06:43:11.4452158Z 🧪 Testing oryx-build-java-project
2025-02-18T06:43:11.7162961Z Operation performed by Microsoft Oryx, https://github.com/Microsoft/Oryx
2025-02-18T06:43:11.7163994Z You can report issues at https://github.com/Microsoft/Oryx/issues
2025-02-18T06:43:11.7164504Z 
2025-02-18T06:43:11.7165103Z Oryx Version: 0.2.0.0+d363b83aad8365d7e91cc246d6dfb3ddad209ddb, Commit: , ReleaseTagName: 
2025-02-18T06:43:11.7165735Z 
2025-02-18T06:43:11.7167170Z Build Operation ID: 03559e34d6abe196
2025-02-18T06:43:11.7167545Z OS Type           : focal-scm
2025-02-18T06:43:11.7167793Z Image Type        : vso-focal
2025-02-18T06:43:11.7167951Z 
2025-02-18T06:43:11.7293139Z Detecting platforms...
2025-02-18T06:43:12.1773019Z Detected following platforms:
2025-02-18T06:43:12.1773544Z   java: 11.0.14
2025-02-18T06:43:12.1811467Z Version '11.0.14' of platform 'java' is not installed. Generating script to install it...
2025-02-18T06:43:12.2623481Z 
2025-02-18T06:43:12.2690670Z 
2025-02-18T06:43:12.2729096Z Source directory     : /workspaces/images/src/universal/test-project/sample/java
2025-02-18T06:43:12.2729945Z Destination directory: /workspaces/images/src/universal/test-project/sample/java
2025-02-18T06:43:12.2730309Z 
2025-02-18T06:43:12.2730313Z 
2025-02-18T06:43:12.2730527Z Downloading and extracting 'java' version '11.0.14' to '/opt/java/11.0.14'...
2025-02-18T06:43:12.2743841Z Detected image debian flavor: focal-scm.
2025-02-18T06:43:15.8144069Z Downloaded in 3 sec(s).
2025-02-18T06:43:15.8149330Z Verifying checksum...
2025-02-18T06:43:15.8203599Z Extracting contents...
2025-02-18T06:43:17.8667912Z performing sha512 checksum for: java...
2025-02-18T06:43:18.4225789Z Done in 6 sec(s).
2025-02-18T06:43:18.4226289Z 
2025-02-18T06:43:18.4245344Z image detector file exists, platform is java..
2025-02-18T06:43:18.4296750Z 
2025-02-18T06:43:18.4297164Z Downloading and extracting 'maven' version '3.8.5' to '/opt/maven/3.8.5'...
2025-02-18T06:43:18.4326078Z Detected image debian flavor: focal-scm.
2025-02-18T06:43:18.7393807Z Downloaded in 0 sec(s).
2025-02-18T06:43:18.7395018Z Verifying checksum...
2025-02-18T06:43:18.7459474Z Extracting contents...
2025-02-18T06:43:18.8158365Z performing sha512 checksum for: maven...
2025-02-18T06:43:18.8453518Z Done in 0 sec(s).
2025-02-18T06:43:18.8453820Z 
2025-02-18T06:43:18.9055857Z Running the command ''
2025-02-18T06:43:18.9056198Z 
2025-02-18T06:43:18.9068769Z 
2025-02-18T06:43:18.9069687Z Removing existing manifest file
2025-02-18T06:43:18.9082983Z Creating a manifest file...
2025-02-18T06:43:18.9089777Z /tmp/BuildScriptGenerator/37fcb599fa1e4aa19c1b987aeb8bb8bc/build.sh: line 340: /workspaces/images/src/universal/test-project/sample/java/oryx-manifest.toml: Permission denied
2025-02-18T06:43:18.9378831Z ❌ oryx-build-java-project check failed.
```

```
2025-02-18T06:43:41.8934016Z 🧪 Testing oryx-build-python
2025-02-18T06:43:42.1640116Z Operation performed by Microsoft Oryx, https://github.com/Microsoft/Oryx
2025-02-18T06:43:42.1640973Z You can report issues at https://github.com/Microsoft/Oryx/issues
2025-02-18T06:43:42.1641430Z 
2025-02-18T06:43:42.1641909Z Oryx Version: 0.2.0.0+d363b83aad8365d7e91cc246d6dfb3ddad209ddb, Commit: , ReleaseTagName: 
2025-02-18T06:43:42.1642485Z 
2025-02-18T06:43:42.1643924Z Build Operation ID: 098e7283c4a60804
2025-02-18T06:43:42.1644631Z OS Type           : focal-scm
2025-02-18T06:43:42.1645037Z Image Type        : vso-focal
2025-02-18T06:43:42.1645300Z 
2025-02-18T06:43:42.1769024Z Detecting platforms...
2025-02-18T06:43:42.5742680Z Detected following platforms:
2025-02-18T06:43:42.5743031Z   python: 3.12.1
2025-02-18T06:43:42.6569110Z 
2025-02-18T06:43:42.6631452Z 
2025-02-18T06:43:42.6638329Z Source directory     : /workspaces/images/src/universal/test-project/sample/python
2025-02-18T06:43:42.6639181Z Destination directory: /workspaces/images/src/universal/test-project/sample/python
2025-02-18T06:43:42.6639808Z 
2025-02-18T06:43:42.7245387Z Python Version: /opt/python/3.12.1/bin/python3.12
2025-02-18T06:43:42.7245982Z Creating directory for command manifest file if it does not exist
2025-02-18T06:43:42.7274614Z Removing existing manifest file
2025-02-18T06:43:42.7293959Z /tmp/BuildScriptGenerator/52d3ea79293e4b93afc2bc9424489b96/build.sh: line 112: /workspaces/images/src/universal/test-project/sample/python/oryx-build-commands.txt: Permission denied
2025-02-18T06:43:42.7575653Z ❌ oryx-build-python check failed.
```

With Regards,
Kaniska

chrmarti replied:
`src/universal/test-project/test.sh` could use `sudo chown -R codespace:codespace /usr/share/dotnet` to update the UID:GIDs there.

Ideally the image and `updateRemoteUserUID` would work together, but maybe that is not so important given that this only comes up now.

Kaniska244 replied:
Hi @chrmarti ,

That works fine for the smoke test step "oryx-install-dotnet-2.1" which was originally failing with the permission denied issue. The only downside is that in this manner during the smoke test updateRemoteUserUID logic comes into picture before starting the container as the host runner UID & GID are different from "codespace" user UID & GID. This adds an extra layer of 1.5 GB on top of the universal image built during the test dueto which the size check parameter of the smoke test starts failing for resultant final image in the smoke test as it executes the logic in 
https://github.com/devcontainers/cli/blob/main/scripts/updateUID.Dockerfile 

```
REPOSITORY                                                                           TAG       IMAGE ID       CREATED         SIZE
vsc-universal-079770f937ff19195472c2be4328d1b0da1657efc298d4bf66edd83df4b0fd5b-uid   latest    f6a22fdd170b   6 minutes ago   12.2GB
vsc-universal-079770f937ff19195472c2be4328d1b0da1657efc298d4bf66edd83df4b0fd5b       latest    7ba076cf1a0d   8 minutes ago   10.7GB
Size of the image sha256:f6a22fdd170b6b8878b9fc922b1b46102b60057a077ce40bcde206c49270f16a: 12249569622 bytes

Threshold is 10737418240 bytes ie 10 GB
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
f6a22fdd170b   6 minutes ago    USER root                                       0B        buildkit.dockerfile.v0
<missing>      6 minutes ago    ARG IMAGE_USER                                  0B        buildkit.dockerfile.v0
### **<missing>      6 minutes ago    RUN |3 REMOTE_USER=codespace NEW_UID=1001 NE…   1.54GB    buildkit.dockerfile.v0**
<missing>      6 minutes ago    SHELL [/bin/sh -c]                              0B        buildkit.dockerfile.v0
<missing>      6 minutes ago    ARG NEW_GID                                     0B        buildkit.dockerfile.v0
<missing>      6 minutes ago    ARG NEW_UID                                     0B        buildkit.dockerfile.v0
<missing>      6 minutes ago    ARG REMOTE_USER                                 0B        buildkit.dockerfile.v0
<missing>      6 minutes ago    USER root                                       0B        buildkit.dockerfile.v0
<missing>      8 minutes ago    ENV CHROME_DEVEL_SANDBOX=/usr/local/sbin/chr…   0B        buildkit.dockerfile.v0
<missing>      8 minutes ago    LABEL devcontainer.metadata=[ {"id":"ghcr.io…   0B        buildkit.dockerfile.v0
<missing>      8 minutes ago    USER root                                       0B        buildkit.dockerfile.v0
<missing>      8 minutes ago    ARG _DEV_CONTAINERS_IMAGE_USER=root             0B        buildkit.dockerfile.v0
<missing>      8 minutes ago    RUN /bin/sh -c cp -ar /tmp/build-features-sr…   1.23GB    buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV PATH=/home/codespace/.dotnet:/home/codes…   0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV JUPYTERLAB_PATH=/home/codespace/.local/b…   0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV GOROOT=/usr/local/go                        0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV RAILS_DEVELOPMENT_HOSTS=.githubpreview.d…   0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV RVM_PATH=/usr/local/rvm                     0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV NVS_HOME=/home/codespace/.nvs               0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV NPM_GLOBAL=/home/codespace/.npm-global      0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV PYTHONIOENCODING=UTF-8                      0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV ORYX_ENV_TYPE=vsonline-present              0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV NUGET_XMLDOC_MODE=skip                      0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1         0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV HUGO_ROOT=/home/codespace/.hugo             0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV MAVEN_ROOT=/home/codespace/.maven           0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV RUBY_ROOT=/home/codespace/.ruby             0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV PYTHON_ROOT=/home/codespace/.python         0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV PHP_ROOT=/home/codespace/.php               0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV NODE_ROOT=/home/codespace/nvm               0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV JAVA_ROOT=/home/codespace/java              0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   ENV RUBY_HOME=/usr/local/rvm/rubies/default     0B        buildkit.dockerfile.v0
<missing>      10 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   1.32GB    buildkit.dockerfile.v0
<missing>      11 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   653B      buildkit.dockerfile.v0
<missing>      11 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   1.19GB    buildkit.dockerfile.v0
<missing>      16 minutes ago   ENV PATH=/usr/local/python/current/bin:/usr/…   0B        buildkit.dockerfile.v0
<missing>      16 minutes ago   ENV PIPX_BIN_DIR=/usr/local/py-utils/bin        0B        buildkit.dockerfile.v0
<missing>      16 minutes ago   ENV PIPX_HOME=/usr/local/py-utils               0B        buildkit.dockerfile.v0
<missing>      16 minutes ago   ENV PYTHON_PATH=/usr/local/python/current       0B        buildkit.dockerfile.v0
<missing>      16 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   820MB     buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV PATH=/usr/local/oryx:/usr/local/go/bin:/…   0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV DEBIAN_FLAVOR=focal-scm                     0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV ORYX_DIR=/usr/local/oryx                    0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV ORYX_PREFER_USER_INSTALLED_SDKS=true        0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV DYNAMIC_INSTALL_ROOT_DIR=/opt               0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV ENABLE_DYNAMIC_INSTALL=true                 0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   ENV ORYX_SDK_STORAGE_BASE_URL=https://oryx-c…   0B        buildkit.dockerfile.v0
<missing>      17 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   41.3MB    buildkit.dockerfile.v0
<missing>      18 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   392MB     buildkit.dockerfile.v0
<missing>      19 minutes ago   ENV PATH=/usr/local/go/bin:/go/bin:/usr/loca…   0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   ENV GOPATH=/go                                  0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   ENV GOROOT=/usr/local/go                        0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   241MB     buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   561MB     buildkit.dockerfile.v0
<missing>      19 minutes ago   ENV DOCKER_BUILDKIT=1                           0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   46.8MB    buildkit.dockerfile.v0
<missing>      20 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   21.1MB    buildkit.dockerfile.v0
<missing>      20 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   6.96MB    buildkit.dockerfile.v0
<missing>      20 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   858MB     buildkit.dockerfile.v0
<missing>      21 minutes ago   ENV PATH=/usr/local/sdkman/bin:/usr/local/sd…   0B        buildkit.dockerfile.v0
<missing>      21 minutes ago   ENV JAVA_HOME=/usr/local/sdkman/candidates/j…   0B        buildkit.dockerfile.v0
<missing>      21 minutes ago   ENV SDKMAN_DIR=/usr/local/sdkman                0B        buildkit.dockerfile.v0
<missing>      21 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   116MB     buildkit.dockerfile.v0
<missing>      24 minutes ago   ENV PATH=/usr/local/rvm/gems/default/bin:/us…   0B        buildkit.dockerfile.v0
<missing>      24 minutes ago   ENV MY_RUBY_HOME=/usr/local/rvm/rubies/defau…   0B        buildkit.dockerfile.v0
<missing>      24 minutes ago   ENV GEM_HOME=/usr/local/rvm/gems/default        0B        buildkit.dockerfile.v0
<missing>      24 minutes ago   ENV GEM_PATH=/usr/local/rvm/gems/default:/us…   0B        buildkit.dockerfile.v0
<missing>      24 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   209MB     buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV PATH=/usr/local/php/current/bin:/opt/con…   0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV PHP_PATH=/usr/local/php/current             0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   218MB     buildkit.dockerfile.v0
<missing>      34 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   419MB     buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV PATH=/opt/conda/bin:/usr/local/nvs:/usr/…   0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV CONDA_SCRIPT=/opt/conda/etc/profile.d/co…   0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV CONDA_DIR=/opt/conda                        0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   1.37MB    buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV PATH=/usr/local/nvs:/usr/local/share/nvm…   0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV NVS_DIR=/usr/local/nvs                      0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   404MB     buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV PATH=/usr/local/share/nvm/current/bin:/u…   0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV NVM_SYMLINK_CURRENT=true                    0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   ENV NVM_DIR=/usr/local/share/nvm                0B        buildkit.dockerfile.v0
<missing>      34 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   50.9MB    buildkit.dockerfile.v0
<missing>      35 minutes ago   ENV PATH=/usr/local/hugo/bin:/usr/local/sbin…   0B        buildkit.dockerfile.v0
<missing>      35 minutes ago   ENV HUGO_DIR=/usr/local/hugo                    0B        buildkit.dockerfile.v0
<missing>      35 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   680MB     buildkit.dockerfile.v0
<missing>      35 minutes ago   ENV DOTNET_USE_POLLING_FILE_WATCHER=true        0B        buildkit.dockerfile.v0
<missing>      35 minutes ago   ENV DOTNET_RUNNING_IN_CONTAINER=true            0B        buildkit.dockerfile.v0
<missing>      35 minutes ago   ENV PATH=/usr/local/sbin:/usr/local/bin:/usr…   0B        buildkit.dockerfile.v0
<missing>      35 minutes ago   ENV DOTNET_ROOT=/usr/share/dotnet               0B        buildkit.dockerfile.v0
<missing>      35 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   180MB     buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN /bin/sh -c cp -ar /tmp/build-features-sr…   125MB     buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN /bin/sh -c echo "_CONTAINER_USER_HOME=$(…   90B       buildkit.dockerfile.v0
<missing>      38 minutes ago   COPY /tmp/build-features/ /tmp/dev-container…   49B       buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN /bin/sh -c mkdir -p /tmp/dev-container-f…   0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   USER root                                       0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN |1 DeveloperBuild= /bin/sh -c if [ -z $D…   0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   ARG DeveloperBuild                              0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   CMD ["sleep" "infinity"]                        0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   VOLUME [/var/lib/docker]                        0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN /bin/sh -c apt-get clean -y && rm -rf /t…   0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN /bin/sh -c apt-get install -yq fish     …   31.6MB    buildkit.dockerfile.v0
<missing>      38 minutes ago   ENV SHELL=/bin/bash DOCKER_BUILDKIT=1           0B        buildkit.dockerfile.v0
<missing>      38 minutes ago   RUN /bin/sh -c apt-get update     && apt-get…   902MB     buildkit.dockerfile.v0
<missing>      39 minutes ago   RUN /bin/sh -c apt-get update     && apt-get…   480MB     buildkit.dockerfile.v0
<missing>      39 minutes ago   ENV LANG=C.UTF-8                                0B        buildkit.dockerfile.v0
<missing>      39 minutes ago   RUN /bin/sh -c apt-get update && export DEBI…   84.7MB    buildkit.dockerfile.v0
<missing>      40 minutes ago   COPY first-run-notice.txt /tmp/scripts/ # bu…   491B      buildkit.dockerfile.v0
<missing>      4 months ago     /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B        
<missing>      4 months ago     /bin/sh -c #(nop) ADD file:7486147a645d8835a…   72.8MB    
<missing>      4 months ago     /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      4 months ago     /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      4 months ago     /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      4 months ago     /bin/sh -c #(nop)  ARG RELEASE                  0B        

🧪 Checking image size of universal :

Image size exceeds the threshold of 10 gb

❌ Image size check failed.
Error: Process completed with exit code 1.
```

With Regards,
Kaniska

chrmarti reviewed (changes requested):
Left a few questions.

chrmarti commented on .github/actions/smoke-test/check-image-size.sh:
```diff
@@ -36,9 +36,14 @@ check_image_size() {
-    IMAGE_ID=$(docker inspect --format='{{.Image}}' "$CONTAINER_ID")
+    if [ $IMAGE == "universal" ]; then
+        #Read the image id of the original image, not the modified image with uid and gid
+        IMAGE_ID=$(docker images | grep -v "uid" | sed '1d' | awk  '{print $3}')
```
This might break depending on the images on the system. Would a constant `mcr.microsoft.com/devcontainers/universal:latest` work? That might already be present at this point?

chrmarti replied:
Makes sense, maybe change the end of `.github/actions/smoke-test/build.sh` to:
```sh
echo "(*) Building image - ${IMAGE}"
image_name="test-image-${IMAGE}"
devcontainer up --image-name ${image_name} --workspace-folder "src/${IMAGE}/"

echo "(*) Starting container - ${IMAGE}"
id_label="test-container=${IMAGE}"
devcontainer up --id-label ${id_label} --workspace-folder "src/${IMAGE}/"
```

That will give us an easy way to refer to the image (the `-uid` image is only built by `devcontainer up`).

Kaniska244 replied:
Hi @chrmarti ,

I have pushed the change as suggested above. Would you kindly have a look & let me know in case of further concerns. 

With Regards,
Kaniska

chrmarti commented on .github/actions/smoke-test/build.sh:
```diff
@@ -7,6 +7,14 @@ export DOCKER_BUILDKIT=1
 id_label="test-container=${IMAGE}"
-devcontainer up --id-label ${id_label} --workspace-folder "src/${IMAGE}/"
+id_image="universal-test-image"
+if [ $IMAGE == "universal" ]; then
```
Looks good. Is there a reason to add the two `if [ $IMAGE == "universal" ]; then` or would this work for all images when using `id_image="${IMAGE}-test-image"`? (But keeping the pre-existing `if` in test.sh.)

Kaniska244 replied:
Hi @chrmarti ,

I have corrected them now to make it more generic. The image size check was originally foreseen for universal image only as I see there were multiple complaints of its huge & ever increasing size reported last year. I am not removing that pre-existing if condition for now. If in future its decided to have this size check enabled for all images, removing that if condition alone should make it work.

Kindly let me know in case of further concern.

With Regards,
Kaniska

chrmarti replied:
Makes sense, thanks.

chrmarti reviewed (approved):
Great, thanks!
